keps/0008-node-heartbeat.md

```diff
@@ -0,0 +1,386 @@
+there will be corresponding `Lease` object with Name equal
+to Node name in the `kube-system` namespace.
+
+We will use CRD (as opposed to built-in API), because:
+
+- CRDs are `the` way to created new APIs
+- even though CRDs aren't very efficient due to lack of protobuf support now:
+    - performance should be acceptable (because Lease object will be small)
+      <br> TODO(wojtek-t): Working on microbenchmark to prove with data.
+    - protobuf support for CRDs is on the roadmap (though not near term)
+
+We will change Kubelet so that:
+
+1. Kubelet is periodically computing NodeStatus every 10s (at it is now), but that will
+   be independent from reporting status
+1. Kubelet is reporting NodeStatus if:
+   - there was a meaningful change in it (initially we can probably assume that every
+     change is meaningful, including e.g. images on the node)
+   - or it didn’t report it over last `node-status-update-period` seconds
+1. Kubelet creates and periodically updates its own Lease object and frequency
+   of those updates is independent from NodeStatus update frequency.
+
+In the meantime, we will change `NodeController` to treat both updates of NodeStatus
+object as well as updates of the new `Lease` object corresponding to a given
+node as healthiness signal from a given Kubelet. This will make it work for both old
+and new Kubelets.
+
+We should also:
+
+1. audit all other existing core controllers to verify if they also don’t require
+   similar changes in their logic ([ttl controller][] being one of the examples)
+1. change controller manager to auto-register that `Lease` CRD
+1. ensure that `Lease` resource is deleted when corresponding node is
+   deleted (probably via owner references)
+1. [out-of-scope] migrate all LeaderElection code to use that CRD
+
+Once all the code changes are done, we will:
+
+1. start updating `Lease` object every 10s by default, at the same time
+   reducing frequency of NodeStatus updates initially to 20s by default.
+   We will reduce it further later.
+   <br> TODO: That still results in higher average QPS. It should be acceptable but
+   needs to be verified.
+1. announce that we are going to reduce frequency of NodeStatus updates further
+   and give people 1-2 releases to switch their code to use `Lease`
+   object (if they relied on frequent NodeStatus changes)
+1. further reduce NodeStatus updates frequency to not less often than once per
+   1 minute.
+   We can’t stop periodically updating NodeStatus as it would be API breaking change,
+   but it’s fine to reduce its frequency (though we should continue writing it at
+   least once per eviction period).
+
+
+To be considered:
+
+1. We may consider reducing frequency of NodeStatus updates to once every 5 minutes
+   (instead of 1 minute). That would help with performance/scalability even more.
+   Caveats:
+   - NodeProblemDetector is currently updating (some) node conditions every 1 minute
```

Comment:
The only reason of updating it periodically is NodeCondition API definition now, which can be changed. NPD can read NodeStatus, and only update upon the changes.

Reply:
I tried to capture that in this point that we generally may change that. If you have any suggestions how to make it more clear, let me know.

Reply:
I think of this as significance filtering. If there is no material change to report, how often should that be confirmed? If there is a material change, how quickly should it be reported?

Reply:
I already tried to answer in the comment above. But the TL;DR is Kubelet:
- if any condition (or any other field of NodeStatus) changed, it will be reported every 10s (as it is now)
- if nothing changed, we guarantee to update the NodeStatus every 1m (or whatever period we decide for)

The logic in NPD should be consistent with that (i.e. compute the NodeStatus every X seconds, report it only if something changed, but not less often than every Y seconds).

Reply:
sgtm